### PR TITLE
Fortunac/chaos caller regs

### PR DESCRIPTION
**Pull request by fortunac**
- Add two new specs:
  1. chaos RAX
  2. chaos all x86_64 caller-saved registers
- Updated documentation for all function specs
- Update the API for creating an environment, so the user can choose which specs that want 
- Allow user to pass in the `--fun-input-regs` flag from the plugin to choose between passing in no inputs vs all input registers at a function call site

### CODE REVIEW

**Comment by fortunac** [comment]
If anyone has better names for  `--fun-input-regs`, I will happily take them.

File: wp/lib/bap_wp/src/environment.ml [hunk 1]
```diff
@@ -205,15 +206,16 @@ let init_loop_unfold (num_unroll : int) : loop_handler =
...
     ~specs:(specs : (Sub.t -> Arch.t -> fun_spec option) list)
     ~default_spec:(default_spec : Sub.t -> Arch.t -> fun_spec)
     ~jmp_spec:(jmp_spec : jmp_spec)
     ~int_spec:(int_spec : int_spec)
+    ~exp_conds:(exp_conds : exp_cond list)
     ~num_loop_unroll:(num_loop_unroll : int)
+    ~arch:(arch : Arch.t)
+    ~freshen_vars:(freshen : bool)
+    ~fun_input_regs:(fun_input_regs : bool)
     (ctx : Z3.context)
```
Comment (codyroux) [inline wp/lib/bap_wp/src/environment.ml]: Any hope of having a simple "default environment" here, like for `mk_ctxt` in Z3?

File: wp/lib/bap_wp/src/precondition.ml [hunk 2]
```diff
@@ -292,17 +292,6 @@ let get_vars (t : Sub.t) : Var.Set.t =
   in
   visitor#visit_sub t Var.Set.empty
 
-(* We are currently filtering out mem from the input variables because it causes
```
Comment (codyroux) [inline wp/lib/bap_wp/src/precondition.ml]: This information might be nice to see in some `info` message, maybe.

File: wp/lib/bap_wp/src/environment.mli [hunk 3]
```diff
@@ -71,18 +71,26 @@ type cond_type = Verify of Constr.goal | Assume of Constr.goal
...
+  :  subs:Bap.Std.Sub.t Bap.Std.Seq.t
   -> specs:(Bap.Std.Sub.t -> Bap.Std.Arch.t -> fun_spec option) list
   -> default_spec:(Bap.Std.Sub.t -> Bap.Std.Arch.t -> fun_spec)
   -> jmp_spec:jmp_spec
   -> int_spec:int_spec
+  -> exp_conds:exp_cond list
   -> num_loop_unroll:int
+  -> arch:Bap.Std.Arch.t
+  -> freshen_vars:bool
+  -> fun_input_regs:bool
```
Comment (codyroux) [inline wp/lib/bap_wp/src/environment.mli]: This is good for now, but in the future we might want to use a more generic way of specifying function summaries.

File: wp/lib/bap_wp/src/environment.ml (hunk 1 above)
Comment (codyroux) [reply wp/lib/bap_wp/src/environment.ml]: Ok I'm an idiot, I forgot we had exactly that in `precondition.ml`.

File: wp/lib/bap_wp/src/precondition.mli [hunk 4]
```diff
@@ -100,10 +92,57 @@ val word_to_z3 : Z3.context -> Bap.Std.Word.t -> Constr.z3_expr
...
+
+(** This spec is used for the functions [__assert_fail] or [__VERIFIER_error]. It
+    returns a precondition of [false]. *)
+val spec_verifier_error : Bap.Std.Sub.t -> Bap.Std.Arch.t -> Env.fun_spec option
+
+(** This spec is used for assumptions made with [__VERIFIER_assume(assumption)].
+    It returns a precondition of [assumption => post]. *)
+val spec_verifier_assume : Bap.Std.Sub.t -> Bap.Std.Arch.t -> Env.fun_spec option
+
+(** This spec is used for functions of [__VERIFIER_nondet_type], which return a
```
Comment (codyroux) [inline wp/lib/bap_wp/src/precondition.mli]: return -> returns

File: wp/lib/bap_wp/src/precondition.mli [hunk 5]
```diff
@@ -100,10 +92,57 @@ val word_to_z3 : Z3.context -> Bap.Std.Word.t -> Constr.z3_expr
...
+val input_regs : Bap.Std.Arch.t -> Bap.Std.Var.t list
+
+(** Obtains the caller saved-registers for a given architecture. *)
+val caller_saved_regs : Bap.Std.Arch.t -> Bap.Std.Var.t list
+
+(** Obtains the callee-saved registers for a given architecture. *)
+val callee_saved_regs : Bap.Std.Arch.t -> Bap.Std.Var.t list
+
+(** This spec is used for the functions [__assert_fail] or [__VERIFIER_error]. It
+    returns a precondition of [false]. *)
```
Comment (codyroux) [inline wp/lib/bap_wp/src/precondition.mli]: "a precondition of..." -> "the precondition..."

File: wp/lib/bap_wp/src/precondition.mli [hunk 6]
```diff
@@ -100,10 +92,57 @@ val word_to_z3 : Z3.context -> Bap.Std.Word.t -> Constr.z3_expr
...
+(** This spec is used to inline a function call. It calls {! visit_sub} on the
+    target function being called. *)
+val spec_inline :
+  Bap.Std.Sub.t Bap.Std.Seq.t -> Bap.Std.Sub.t -> Bap.Std.Arch.t -> Env.fun_spec option
+
 (** The default spec used when mapping subroutines to their preconditions. This
     spec sets the constraint representing the subroutine being called to true, and
-    updates the value of the stack pointer on the return, using a heuristic based on
-    the increments of [SP] before the return in the last block of the subroutine. *)
+    in x86 architectures, increments the value of the stack pointer on the return
```
Comment (codyroux) [inline wp/lib/bap_wp/src/precondition.mli]: "on the return" -> "on return"

File: wp/lib/bap_wp/src/precondition.mli [hunk 7]
```diff
@@ -127,19 +166,26 @@ val non_null_assert : Env.exp_cond
     We use the default value [!num_unroll = 5]. *)
 val num_unroll : int ref
 
-
-(** Creates an environment that inlines function calls not defined in
-    cbat.h.  It contains a default list of jump specs and interrupt
-    specs, and an empty list of {!Environment.exp_cond}s. Takes in a sequence of subs
-    of the program, and optionally the number of times to unroll a loop,
-    a list of {!Environment.exp_cond}s, and the architecture of the binary. *)
+(** Creates an environment with an empty sequence of subroutines to initialize
```
Comment (codyroux) [inline wp/lib/bap_wp/src/precondition.mli]: Maybe a bulleted list here for clarity?

File: wp/plugin/wp.ml [hunk 8]
```diff
@@ -228,6 +232,12 @@ module Cmdline = struct
             at each jump in the path. The path contains information about whether \
             a jump has been taken and the address of the jump if found."
 
+  let fun_input_regs = param bool "fun-input-regs" ~default:true
+      ~doc:"If set, at a function call site, uses all possible input registers \
+            as arguments to a function predicate generated for an output register \
```
Comment (codyroux) [inline wp/plugin/wp.ml]: "function predicate [...] that holds the result of the function call" might be better called a "function symbol [...] that represents the result of the function call"? YMMV

**Comment by codyroux** [comment]
Looks fantastic! Just nitpicks, as usual.

File: wp/lib/bap_wp/src/environment.ml (hunk 1 above)
Comment (philzook58) [reply wp/lib/bap_wp/src/environment.ml]: Is there a reason to have the parameter named freshen_vars and the variable freshen?

**Comment by philzook58** [comment]
Is the README out of date? Maybe we should remove the arguments section of the README and just says how to print out the options?

**Comment by philzook58** [comment]
Looks good otherwise! Builds and tests pass on my machine

File: wp/lib/bap_wp/src/environment.ml (hunk 1 above)
Comment (fortunac) [reply wp/lib/bap_wp/src/environment.ml]: I think we should be consistent. I'll change it to `freshen_vars`

**Pull request merged**